openjdk/src/main/java/org/conscrypt/NativeCryptoJni.java

```diff
@@ -16,126 +16,112 @@
  * build.
  */
 final class NativeCryptoJni {
-    private static final String LIB_NAME = "conscrypt_openjdk_jni";
-    private static final String UNKNOWN = "unknown";
-    private static final String LINUX = "linux";
-
-    public static void init() {
-        String os = normalizeOs(System.getProperty("os.name", ""));
-        String arch = normalizeArch(System.getProperty("os.arch", ""));
-
-        // First, try loading the platform-specific library. Platform-specific
-        // libraries will be available if using a tcnative uber jar.
-        //
-        // TODO(davidben): Pick one name for the library and stick with it.
-        String platformSpecific = LIB_NAME + "-" + os + '-' + arch;
-        NativeLibraryLoader.loadFirstAvailable(
-                NativeCrypto.class.getClassLoader(), platformSpecific, LIB_NAME);
-    }
-
-    private NativeCryptoJni() {}
+    private static final String STATIC_LIB_NAME = "conscrypt";
+    private static final String DYNAMIC_LIB_NAME_PREFIX = "conscrypt_openjdk_jni";
 
     /**
-     * Normalizes the os.name value into the value used by the Maven os plugin
-     * (https://github.com/trustin/os-maven-plugin). This plugin is used to generate
-     * platform-specific
-     * classifiers for artifacts.
+     * Attempts to load the shared JNI library. First try loading the platform-specific library
+     * name (e.g. conscrypt_openjdk_jni-linux-x86_64). If not found, try the static library name.
+     *
+     * The static library name is needed in order to support Java 8 static linking
+     * (http://openjdk.java.net/jeps/178), where the library name is used to invoke a
+     * library-specific load method (i.e. {@code JNI_OnLoad_conscrypt}).
+     *
+     * @throws UnsatisfiedLinkError if the library failed to load.
      */
-    private static String normalizeOs(String value) {
-        value = normalize(value);
-        if (value.startsWith("aix")) {
-            return "aix";
+    static void init() throws UnsatisfiedLinkError {
+        List<LoadResult> results = new ArrayList<LoadResult>();
+        if (!NativeLibraryLoader.loadFirstAvailable(
+                    classLoader(), results, platformLibName(), STATIC_LIB_NAME)) {
+            logResults(results);
+            throwBestError(results);
         }
-        if (value.startsWith("hpux")) {
-            return "hpux";
+    }
+
+    private NativeCryptoJni() {}
+
+    private static void logResults(List<LoadResult> results) {
+        for (LoadResult result : results) {
+            result.log();
         }
-        if (value.startsWith("os400")) {
-            // Avoid the names such as os4000
-            if (value.length() <= 5 || !Character.isDigit(value.charAt(5))) {
-                return "os400";
+    }
+
+    private static void throwBestError(List<LoadResult> results) {
+        Collections.sort(results, ErrorComparator.INSTANCE);
+
+        UnsatisfiedLinkError bestError = null;
```

Comment:
TIOLI: Is this simpler with 

```
UnsatisfiedLinkError bestError = results.get(0);
for (LoadResult result : results.subList(1)) {
  // Suppress all of the other errors, so that they're available to the caller if
  // desired. Note: suppression is only supported on Java 7+.
  Platform.addSuppressed(bestError, result.error);
}
```